Cesium3DTilesSelection/src/CompositeContent.cpp

```diff
@@ -97,69 +106,79 @@ CompositeContent::load(const TileContentLoadInput& input) {
         "Composite tile byteLength is {} but only {} bytes are available.",
         pHeader->byteLength,
         data.size());
-    return nullptr;
+    return asyncSystem.createResolvedFuture(
+        std::unique_ptr<TileContentLoadResult>(nullptr));
   }
 
-  std::vector<std::unique_ptr<TileContentLoadResult>> innerTiles;
-
+  std::vector<CesiumAsync::Future<std::unique_ptr<TileContentLoadResult>>>
+      innerTiles;
   uint32_t pos = sizeof(CmptHeader);
 
-  for (uint32_t i = 0; i < pHeader->tilesLength; ++i) {
+  for (uint32_t i = 0; i < pHeader->tilesLength && pos < pHeader->byteLength;
+       ++i) {
     if (pos + sizeof(InnerHeader) > pHeader->byteLength) {
       SPDLOG_LOGGER_WARN(
           pLogger,
           "Composite tile ends before all embedded tiles could be read.");
+      pos = static_cast<uint32_t>(pHeader->byteLength);
       break;
     }
-
     const InnerHeader* pInner =
         reinterpret_cast<const InnerHeader*>(data.data() + pos);
     if (pos + pInner->byteLength > pHeader->byteLength) {
       SPDLOG_LOGGER_WARN(
           pLogger,
           "Composite tile ends before all embedded tiles could be read.");
+      pos = static_cast<uint32_t>(pHeader->byteLength);
       break;
     }
-
     gsl::span<const std::byte> innerData(data.data() + pos, pInner->byteLength);
 
-    std::unique_ptr<TileContentLoadResult> pInnerLoadResult =
-        TileContentFactory::createContent(derive(input, innerData));
-
-    if (pInnerLoadResult) {
-      innerTiles.emplace_back(std::move(pInnerLoadResult));
-    }
-
     pos += pInner->byteLength;
-  }
 
-  if (innerTiles.empty()) {
-    if (pHeader->tilesLength > 0) {
-      SPDLOG_LOGGER_WARN(
-          pLogger,
-          "Composite tile does not contain any loadable inner tiles.");
-    }
-    return nullptr;
-  }
-  if (innerTiles.size() == 1) {
-    return std::move(innerTiles[0]);
-  }
-
-  std::unique_ptr<TileContentLoadResult> pResult = std::move(innerTiles[0]);
-
-  for (size_t i = 1; i < innerTiles.size(); ++i) {
-    if (!innerTiles[i]->model) {
-      continue;
-    }
-
-    if (pResult->model) {
-      pResult->model.value().merge(std::move(innerTiles[i]->model.value()));
-    } else {
-      pResult->model = std::move(innerTiles[i]->model);
-    }
+    innerTiles.push_back(TileContentFactory::createContent(
+        asyncSystem,
+        pAssetAccessor,
+        requestHeaders,
+        derive(input, innerData)));
   }
 
-  return pResult;
+  return asyncSystem.all(std::move(innerTiles))
+      .thenInMainThread(
```

Comment:
This doesn't need to be done in the main thread, does it? `thenInWorkerThread` would be better, I think.